File: analysis/preprocessing/prepare_sc_ZhangTNBC2021_v1.R
#### -------------------------------------------------------------------------------
#### created on 12 apr 2023, 04:40pm
#### author: dhrubas2
#### -------------------------------------------------------------------------------

.wpath. <- "/Users/dhrubas2/OneDrive - National Institutes of Health/Projects/TMEcontribution/analysis/submission/Code/analysis/"
.mpath. <- "miscellaneous/r/miscellaneous.R"
setwd(.wpath.)                                                                      # current path
source(.mpath.)

library(Matrix)

fcat <- function(...) cat(paste0(glue::glue(...), "\n"))                            # f-string print akin to python

cat("\014")                                                                         # clears console


#### -------------------------------------------------------------------------------

## get data & annotations.
data.paths <- c("../data/SC_data/ZhangTNBC2021/", 
                "../data/TransNEO/use_data/")
data.files <- c("TNBC_scExp_resp_pre_ZhangEtAl2021.RDS", 
                "TNBC_scData_ZhangEtAl2021.xlsx", 
                "gene_length_ensembl.grch37.87_SRD_17Mar2022.txt")

tnbc.data  <- readRDS(paste0(data.paths[1], data.files[1]))

## use TransNEO annotations (hg19) to filter protein-coding genes.
gene.annot <- read.table(paste0(data.paths[2], data.files[3]), sep = "\t", 
                         header = T, as.is = T)
gene.annot.pc <- gene.annot %>% filter(gene_biotype == "protein_coding") %>% 
    (function(df) df[!duplicated(df$gene_name), ]) %>% `rownames<-`(.$gene_name)

## patient/sample lists by treatment + sample type.
sample.type <- "tissue"                                                             # use "tissue" samples (other option: "blood")
sample.type <- sample.type %>% substr(1, 1) %>% `names<-`(sample.type)
sc.samples  <- tnbc.data$patients$Treatment %>% unique %>% 
    sapply(simplify = F, function(trt){
        tnbc.data$patients %>% filter(Treatment == trt, Origin == sample.type) %>% 
            .$Sample.id
    })


## rename cell types to match with TransNEO.
## B cell: B-cells, T cell: T-cells, Myeloid cell: Myeloid, ILC cell: ILC
cells.new <- tnbc.data$cells %>% sapply(function(ctp){
    strsplit(ctp, split = " ")[[1]] %>% (function(x){ 
        ifelse((x[1] %>% str_length) == 1, 
               yes = paste(x[1], "cells", sep = "-"),                               # B cell / T cell
               no = x[1])
    })
})


#### -------------------------------------------------------------------------------

## relevant functions.
## rename cell types to match with deconvolved data.
rename.cell.type <- function(ctp.list){
    ctp.list.new <- ctp.list
    for (ctp in cells.new %>% names){
        ctp.list.new <- ctp.list.new %>% 
            gsub(pattern = ctp, replacement = cells.new[ctp])
    }
    ctp.list.new
}

## SC data is from 10X - in log(1 + x) format, where sum(x) = 1e4.
## reverse the log-transform & multiply by 100 for TPM equivalence.
tpm.from.10x <- function(expr.10x){
    expr.10x.tpm <- ((exp(expr.10x) - 1) * 100) %>% as.matrix %>% as.data.frame
    expr.10x.tpm
}


#### -------------------------------------------------------------------------------

## get chemotherapy data.
treat <- "Chemo"

samples.cm <- sc.samples[[treat]]
annot.cm   <- tnbc.data$annot %>% 
    filter(Treatment == treat, Origin == sample.type) %>% 
    mutate(Cell.type = .$Cluster_org %>% rename.cell.type)

pb <- ProgressBar(N = tnbc.data$cells %>% length)
sc.exp.cm <- tnbc.data$cells %>% sapply(simplify = F, function(ctp){
    pb$tick()
    
    ## get relevant cell ids.
    annot.ctp  <- annot.cm %>% filter(Cluster_org == ctp)
    cells.ctp  <- samples.cm %>% lapply(function(smpl){
        annot.ctp %>% filter(Sample.id == smpl) %>% .$Cell.id                       # cell ids per patient sample
    }) %>% Reduce(f = union)
    
    ## keep protein-coding genes only.
    sc.exp.ctp <- tnbc.data$tissue.norm[[ctp]]                                      # use normalized sc exp
    genes.ctp  <- sc.exp.ctp %>% rownames %>% intersect(., gene.annot.pc$gene_name)
    sc.exp.ctp <- sc.exp.ctp[genes.ctp, cells.ctp] %>% tpm.from.10x                 # convert to tpm
    sc.exp.ctp
})
names(sc.exp.cm) <- cells.new


#### -------------------------------------------------------------------------------

## get immunotherapy data.
treat <- "Anti-PD-L1+Chemo"

samples.im <- sc.samples[[treat]]
annot.im   <- tnbc.data$annot %>% 
    filter(Treatment == treat, Origin == sample.type) %>% 
    mutate(Cell.type = .$Cluster_org %>% rename.cell.type)

pb <- ProgressBar(N = tnbc.data$cells %>% length)
sc.exp.im <- tnbc.data$cells %>% sapply(simplify = F, function(ctp){
    pb$tick()
    
    ## get relevant cell ids.
    annot.ctp  <- annot.im %>% filter(Cluster_org == ctp)
    cells.ctp  <- samples.im %>% lapply(function(smpl){
        annot.ctp %>% filter(Sample.id == smpl) %>% .$Cell.id                       # cell ids per patient sample
    }) %>% Reduce(f = union)
    
    ## keep protein-coding genes only.
    sc.exp.ctp <- tnbc.data$tissue.norm[[ctp]]                                      # use normalized sc exp
    genes.ctp  <- sc.exp.ctp %>% rownames %>% intersect(., gene.annot.pc$gene_name)
    sc.exp.ctp <- sc.exp.ctp[genes.ctp, cells.ctp] %>% tpm.from.10x
    sc.exp.ctp
})
names(sc.exp.im) <- cells.new


#### -------------------------------------------------------------------------------

## save data.
svdat <- F                                                                          # set T to save data 
if (svdat){
    ## save chemo data.
    out.file <- glue("TNBC_scExp_{sample.type %>% names}_pre_Chemo_ZhangEtAl2021.tsv")
    pb <- ProgressBar(N = cells.new %>% length)
    cells.new %>% plyr::l_ply(function(ctp){
        pb$tick()
        out.file.ctp <- out.file %>% 
            gsub(pattern = ".tsv", replacement = glue("_{ctp}.tsv"))
        write.table(sc.exp.cm[[ctp]], file = paste0(data.paths[1], out.file.ctp), 
                    sep = "\t", col.names = T, row.names = T)
    })
    # WriteXlsx(sc.exp.cm, file.name = paste0(data.paths[1], out.file))
    
    ## save anti-PD-L1+chemo data.
    out.file <- glue("TNBC_scExp_{sample.type %>% names}_pre_Anti-PD-L1_Chemo_ZhangEtAl2021.tsv")
    pb <- ProgressBar(N = cells.new %>% length)
    cells.new %>% plyr::l_ply(function(ctp){
        pb$tick()
        out.file.ctp <- out.file %>% 
            gsub(pattern = ".tsv", replacement = glue("_{ctp}.tsv"))
        write.table(sc.exp.im[[ctp]], file = paste0(data.paths[1], out.file.ctp), 
                    sep = "\t", col.names = T, row.names = T)
    })
    # WriteXlsx(sc.exp.im, file.name = paste0(data.paths[1], out.file))
    
    ## save annotations.
    out.file <- glue("TNBC_scAnnot_{sample.type %>% names}_pre_ZhangEtAl2021.xlsx")
    out.obj  <- list("Chemo" = annot.cm, "Anti-PD-L1_Chemo" = annot.im)
    WriteXlsx(out.obj, file.name = paste0(data.paths[1], out.file), 
              col.names = T, row.names = F)
}

File: analysis/preprocessing/prepare_sc_ZhangTNBC2021_v1.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Mon Apr 17 13:08:35 2023

@author: dhrubas2
"""

## set up necessary directories/paths.
_wpath_ = "/Users/dhrubas2/OneDrive - National Institutes of Health/Projects/TMEcontribution/analysis/submission/Code/analysis/"
_mpath_ = "miscellaneous/py/"

## load necessary packages.
import os, sys
sys.path.append(_wpath_);       os.chdir(_wpath_)                              # current path
if _mpath_ not in sys.path:
    sys.path.append(_mpath_)                                                   # to load miscellaneous

import numpy as np, pandas as pd, pickle
from miscellaneous import list_files, tic
from tqdm import tqdm


#%% functions.

def read_data(file, sheet = None, sep = "\t", **kwarg):
    try:
        data = pd.read_excel(file, sheet_name = sheet, header = 0, 
                             index_col = None, **kwarg)
    except:
        data = pd.read_table(file, sep = sep, header = 0, index_col = 0, 
                             **kwarg)
    
    return data


## isolate cells per cell type per patient/sample.
def get_cell_ids(clin, keep_ = None):
    if keep_ is None:
        keep_ = ["Sample.id", "Patient.id"]
    
    cell_ids = {ctp_: clin[clin["Cell.type"] == ctp_][keep_].reset_index() \
                for ctp_ in clin["Cell.type"].unique()}
    
    return cell_ids


## compute pseudo-bulk expression across all available cell types.
def get_pseudo_bulk(exp_sc, sample_list):
    exp_pblk = { }
    for smpl_ in tqdm(sample_list):
        exp_ctps_ = pd.concat([
            exp_.iloc[:, exp_.columns.map(lambda x: smpl_ in x)] \
                for exp_ in exp_sc.values()], axis = 1)
        exp_pblk[smpl_] = exp_ctps_.mean(axis = 1)
    
    exp_pblk = pd.DataFrame(exp_pblk)
    
    return exp_pblk
    

#%% read data.

data_path  = "../data/SC_data/ZhangTNBC2021/"
data_files = ["TNBC_scAnnot_tissue_pre_ZhangEtAl2021.xlsx"] + \
    list_files(data_path, pattern = "Chemo") + \
        list_files(data_path, pattern = "Anti-PD-L1_Chemo")


## available cell types + treatments.
cell_types = ["B-cells", "ILC", "Myeloid", "T-cells"]
treatments = ["Chemo", "Anti-PD-L1_Chemo"]

## get data.
clin_info = read_data(data_path + data_files[0], sheet = None)

_tic = tic()
exp_sc_chemo, exp_sc_chemo_immuno = { }, { }
for ctp_ in tqdm(cell_types):
    data_files += [
        f"TNBC_scExp_tissue_pre_Chemo_ZhangEtAl2021_{ctp_}.tsv", 
        f"TNBC_scExp_tissue_pre_Anti-PD-L1_Chemo_ZhangEtAl2021_{ctp_}.tsv"]
        
    exp_sc_chemo[ctp_]        = read_data(data_path + data_files[-2])
    exp_sc_chemo_immuno[ctp_] = read_data(data_path + data_files[-1])

del ctp_
_tic.toc()


#%% prepare response & pseudobulk expression.

## chemo.
clin_info_cm = clin_info["Chemo"].dropna(
    subset = "Efficacy", how = "any").set_index("Cell.id")
clin_info_cm["Response"] = clin_info_cm["Efficacy"].eq("PR").astype(int)       # efficacy is in RECIST scale (no CR) => PR == R
cell_ids_cm  = get_cell_ids(clin_info_cm)
resp_PR_cm   = clin_info_cm[["Patient.id", "Sample.id", "Response"]].copy()
exp_sc_cm    = {ctp_: exp_[cell_ids_cm[ctp_]["Cell.id"]] \
                for ctp_, exp_ in exp_sc_chemo.items()}
exp_sc_cm["PseudoBulk"] = get_pseudo_bulk(
    exp_sc_cm, sample_list = clin_info_cm["Sample.id"].unique())

data_cm     = {"exp": exp_sc_cm, "resp": resp_PR_cm, "cells": cell_ids_cm, 
               "clin": clin_info_cm}


## chemo + immuno.
clin_info_im = clin_info["Anti-PD-L1_Chemo"].dropna(
    subset = "Efficacy", how = "any").set_index("Cell.id")
clin_info_im["Response"] = clin_info_im["Efficacy"].eq("PR").astype(int)       # efficacy is in RECIST scale (no CR) => PR == R
cell_ids_im  = get_cell_ids(clin_info_im)
resp_PR_im   = clin_info_im[["Patient.id", "Sample.id", "Response"]].copy()
exp_sc_im    = {ctp_: exp_[cell_ids_im[ctp_]["Cell.id"]] \
                for ctp_, exp_ in exp_sc_chemo_immuno.items()}
exp_sc_im["PseudoBulk"] = get_pseudo_bulk(
    exp_sc_im, sample_list = clin_info_im["Sample.id"].unique())

data_im      = {"exp": exp_sc_im, "resp": resp_PR_im, "cells": cell_ids_im, 
                "clin": clin_info_im}


## all.
clin_info_all = pd.concat([clin_info_cm, clin_info_im], axis = 0)
cell_ids_all  = {ctp_: pd.concat([cell_ids_cm[ctp_], cell_ids_im[ctp_]], 
                                 axis = 0) for ctp_ in tqdm(cell_types)}
resp_PR_all   = clin_info_all[["Patient.id", "Sample.id", "Response"]].copy()
exp_sc_all    = {ctp_: pd.concat([exp_sc_cm[ctp_], exp_sc_im[ctp_]], axis = 1) \
                 for ctp_ in tqdm(np.append(cell_types, "PseudoBulk"))}

data_all      = {"exp": exp_sc_all, "resp": resp_PR_all, "cells": cell_ids_all, 
                 "clin": clin_info_all}


#%% save data.

svdat = False                                                                  # set True to save data 

if svdat:
    out_path = "../data/SC_data/ZhangTNBC2021/validation/"
    os.makedirs(out_path, exist_ok = True)                                     # creates output dir if doesn't exist
    
    ## save chemo data.
    print("\nsaving data for Chemo...", end = " ")
    out_file = "tnbc_sc_data_chemo_v2.pkl"
    with open(out_path + out_file, "wb") as file:
        pickle.dump(data_cm, file)
    print(f"done!\ndata is in = '{out_path + out_file}'")
    
    ## save chemo + immuno data.
    print("\nsaving data for Chemo + Immuno...", end = " ")
    out_file = "tnbc_sc_data_chemo_immuno_v2.pkl"
    with open(out_path + out_file, "wb") as file:
        pickle.dump(data_im, file)
    print(f"done!\ndata is in = '{out_path + out_file}'")
    
    ## save both chemo & chemo + immuno data.
    print("\nsaving data for both treatments...", end = " ")
    out_file = "tnbc_sc_data_all_v2.pkl"
    with open(out_path + out_file, "wb") as file:
        pickle.dump(data_all, file)
    print(f"done!\ndata is in = '{out_path + out_file}'")

